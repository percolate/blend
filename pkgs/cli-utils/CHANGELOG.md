# Change Log

All notable changes to this project will be documented in this file.
See [Conventional Commits](https://conventionalcommits.org) for commit guidelines.

## 0.0.1 (2019-08-28)


### Bug Fixes

* expose IParallelizeOpts type def ([5695eb1](https://github.com/percolate/blend/tree/master/pkgs/core/commit/5695eb1))
* verbose execSync duplicates error message ([e808ebe](https://github.com/percolate/blend/tree/master/pkgs/core/commit/e808ebe))
