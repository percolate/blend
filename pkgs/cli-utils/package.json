{
  "name": "@percolate/cli-utils",
  "version": "0.1.23",
  "description": "Percolate CLI utils",
  "license": "CPAL-1.0",
  "repository": "https://github.com/percolate/blend/tree/master/pkgs/core",
  "main": "./dist/index.js",
  "types": "./dist/index.d.ts",
  "files": [
    "dist",
    "src"
  ],
  "scripts": {
    "build": "tsc --build tsconfig.json",
    "watch": "tsc --build tsconfig.json --watch --preserveWatchOutput"
  },
  "dependencies": {
    "chalk": "3.0.0",
    "fs-readdir-recursive": "1.1.0",
    "micromatch": "4.0.6"
  },
  "publishConfig": {
    "access": "public"
  },
  "volta": {
    "extends": "../../package.json"
  }
}
