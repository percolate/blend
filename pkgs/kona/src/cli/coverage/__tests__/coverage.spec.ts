import { cleanExit, execSync } from '@percolate/cli-utils'
import * as fs from 'fs'
import * as path from 'path'

// prevent force exiting of tests
// prevent fetching remote
jest.mock('@percolate/cli-utils', () => {
    const cliUtils2 = jest.requireActual('@percolate/cli-utils')
    return {
        ...cliUtils2,
        cleanExit: jest.fn(),
        forceExit: jest.fn(),
        git: {
            ...cliUtils2.git,
            fetchLatestMaster: () => {},
        },
    }
})

// set the fixture folder as project root
const fixturePath = path.resolve(__dirname, '__fixtures__', 'coverage')
process.chdir(fixturePath)
jest.mock('../../../root.ts', () => ({
    root: (...resolvePaths: string[]) => {
        const path2 = jest.requireActual('path')
        const root = path2.resolve(__dirname, '__fixtures__', 'coverage')
        return resolvePaths.length ? `${path2.resolve(root, ...resolvePaths)}` : root
    },
}))

let mockOutput: string[] = []
console.log = (...args: string[]) => {
    mockOutput.push(args[0])
}

// eslint-disable-next-line import/no-commonjs
const { coverageCmd } = require('../coverage')

const branchExists = (branch: string) => {
    let exists: boolean
    exists = !!execSync(`git branch --list ${branch}`, {
        onError: () => {
            exists = false
        },
    })
    return exists
}

const commitChanges = () => {
    execSync('git add -A')
    execSync('git commit -m "changes"')
}

const resetBranches = () => {
    if (!branchExists('feature')) return
    execSync('git checkout master')
    execSync('git branch -D feature')
    execSync('git stash')
}

const removeVersionControl = () => {
    if (!fs.existsSync(`${fixturePath}/.git`)) return
    execSync(`rm -rf ${fixturePath}/.git`)
}

const resetVersionControl = () => {
    removeVersionControl()
    execSync('git init')
    commitChanges()
}

const cmd = async () => {
    const res = await coverageCmd.handler({ silent: true } as any)
    // eslint-disable-next-line no-control-regex
    return res ? res.replace(/\x1B[[(?);]{0,2}(;?\d)*./g, '') : res
}

// eslint-disable-next-line no-control-regex
const getOutput = () => mockOutput.join('\n').replace(/\x1B[[(?);]{0,2}(;?\d)*./g, '')

beforeAll(() => {
    resetVersionControl()
})

beforeEach(() => {
    mockOutput = []
    resetBranches()
    execSync('git branch feature')
    execSync('git checkout feature')
    execSync(`mkdir ${fixturePath}/tmp`)
    execSync(`mkdir ${fixturePath}/tmp/reports`)
})

afterEach(() => {
    jest.clearAllMocks()
    resetBranches()
    execSync(`rm -rf ${fixturePath}/tmp`)
})

afterAll(() => {
    removeVersionControl()
})

it('should not run on master', () => {
    execSync('git checkout master')
    const cleanExitMock = cleanExit as jest.MockedFunction<any>
    cmd()
    expect(cleanExitMock.mock.calls[0][0]).toBe("Skipped: 'master' branch")
})

it('should not report if nothing changes', async () => {
    const coverage = `
TN:
SF:util.ts
FN:1,(anonymous_0)
FN:2,(anonymous_1)
FNF:2
FNH:1
FNDA:1,(anonymous_0)
FNDA:0,(anonymous_1)
DA:1,1
DA:2,1
LF:2
LH:2
BRF:0
BRH:0
end_of_record    
`
    fs.writeFileSync(`${fixturePath}/tmp/reports/lcov.info`, coverage, 'utf8')
    await cmd()
    expect(getOutput()).toMatchSnapshot()
})

it('should not report deletions', async () => {
    const changes = `export const add = (num1: number, num2: number) => num1 + num2
`

    const coverage = `
TN:
SF:util.ts
FN:1,(anonymous_0)
FN:2,(anonymous_1)
FNF:2
FNH:1
FNDA:1,(anonymous_0)
FNDA:0,(anonymous_1)
DA:1,1
DA:2,1
LF:2
LH:2
BRF:0
BRH:0
end_of_record    
`

    fs.writeFileSync(`${fixturePath}/util.ts`, changes, 'utf8')
    fs.writeFileSync(`${fixturePath}/tmp/reports/lcov.info`, coverage, 'utf8')
    commitChanges()
    await cmd()
    expect(getOutput()).toMatchSnapshot()
})

it('should report new lines', async () => {
    const changes = `export const add = (num1: number, num2: number) => num1 + num2
export const subtract = (num1: number, num2: number) => num1 - num2
export const math = { add, subtract }
`

    const coverage = `
TN:
SF:util.ts
FN:1,(anonymous_0)
FN:2,(anonymous_1)
FNF:2
FNH:1
FNDA:1,(anonymous_0)
FNDA:0,(anonymous_1)
DA:1,1
DA:2,1
LF:2
LH:2
BRF:0
BRH:0
end_of_record    
`

    fs.writeFileSync(`${fixturePath}/util.ts`, changes, 'utf8')
    fs.writeFileSync(`${fixturePath}/tmp/reports/lcov.info`, coverage, 'utf8')
    commitChanges()
    await cmd()
    expect(getOutput()).toMatchSnapshot()
})

it('should report new functions', async () => {
    const changes = `export const add = (num1: number, num2: number) => num1 + num2
export const subtract = (num1: number, num2: number) => num1 - num2
export const divide = (num1: number, num2: number) => num1 / num2
`

    const coverage = `
TN:
SF:util.ts
FN:1,(anonymous_0)
FN:2,(anonymous_1)
FN:3,(anonymous_2)
FNF:3
FNH:1
FNDA:1,(anonymous_0)
FNDA:0,(anonymous_1)
FNDA:0,(anonymous_2)
DA:1,1
DA:2,1
DA:3,1
LF:3
LH:3
BRF:0
BRH:0
end_of_record
`

    fs.writeFileSync(`${fixturePath}/util.ts`, changes, 'utf8')
    fs.writeFileSync(`${fixturePath}/tmp/reports/lcov.info`, coverage, 'utf8')
    commitChanges()
    await cmd()
    expect(getOutput()).toMatchSnapshot()
})

it('should report new branches', async () => {
    const changes = `export const add = (num1: number, num2: number) => num1 ? num1 + num2 : num1
export const subtract = (num1: number, num2: number) => num1 - num2
`
    const coverage = `
TN:
SF:util.ts
FN:1,(anonymous_0)
FN:2,(anonymous_1)
FNF:2
FNH:1
FNDA:1,(anonymous_0)
FNDA:0,(anonymous_1)
DA:1,1
DA:2,1
LF:2
LH:2
BRDA:1,0,0,1
BRDA:1,0,1,0
BRF:2
BRH:1
end_of_record
`

    fs.writeFileSync(`${fixturePath}/util.ts`, changes, 'utf8')
    fs.writeFileSync(`${fixturePath}/tmp/reports/lcov.info`, coverage, 'utf8')
    commitChanges()
    await cmd()
    expect(getOutput()).toMatchSnapshot()
})

it('should report all three', async () => {
    const changes = `export const add = (num1: number, num2: number) => num1 ? num1 + num2 : num1
export const subtract = (num1: number, num2: number) => num1 - num2
export const divide = (num1: number, num2: number) => num1 / num2
export const math = { add, divide, subtract }
`

    const coverage = `
TN:
SF:util.ts
FN:1,(anonymous_0)
FN:2,(anonymous_1)
FN:3,(anonymous_2)
FNF:3
FNH:1
FNDA:1,(anonymous_0)
FNDA:0,(anonymous_1)
FNDA:0,(anonymous_2)
DA:1,1
DA:2,1
DA:3,1
DA:4,1
LF:4
LH:4
BRDA:1,0,0,1
BRDA:1,0,1,0
BRF:2
BRH:1
end_of_record
`

    fs.writeFileSync(`${fixturePath}/util.ts`, changes, 'utf8')
    fs.writeFileSync(`${fixturePath}/tmp/reports/lcov.info`, coverage, 'utf8')
    commitChanges()
    await cmd()
    expect(getOutput()).toMatchSnapshot()
})

it('should report changes across multiple files', async () => {
    const changes = `export const add = (num1: number, num2: number) => num1 ? num1 + num2 : num1
export const subtract = (num1: number, num2: number) => num1 - num2
export const divide = (num1: number, num2: number) => num1 / num2
export const math = { add, divide, subtract }
`

    const coverage = `
TN:
SF:util.ts
FN:1,(anonymous_0)
FN:2,(anonymous_1)
FN:3,(anonymous_2)
FNF:3
FNH:1
FNDA:1,(anonymous_0)
FNDA:0,(anonymous_1)
FNDA:0,(anonymous_2)
DA:1,1
DA:2,1
DA:3,1
DA:4,1
LF:4
LH:4
BRDA:1,0,0,1
BRDA:1,0,1,0
BRF:2
BRH:1
end_of_record
TN:
SF:util1.ts
FN:1,(anonymous_0)
FN:2,(anonymous_1)
FN:3,(anonymous_2)
FNF:3
FNH:0
FNDA:0,(anonymous_0)
FNDA:0,(anonymous_1)
FNDA:0,(anonymous_2)
DA:1,0
DA:2,0
DA:3,0
DA:4,0
LF:4
LH:0
BRDA:1,0,0,0
BRDA:1,0,1,0
BRF:2
BRH:0
end_of_record
`

    fs.writeFileSync(`${fixturePath}/util.ts`, changes, 'utf8')
    execSync(`touch ${fixturePath}/util1.ts`)
    fs.writeFileSync(`${fixturePath}/util1.ts`, changes, 'utf8')
    fs.writeFileSync(`${fixturePath}/tmp/reports/lcov.info`, coverage, 'utf8')
    commitChanges()
    await cmd()
    expect(getOutput()).toMatchSnapshot()
})
