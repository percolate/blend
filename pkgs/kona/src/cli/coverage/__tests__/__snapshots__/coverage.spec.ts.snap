// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should not report deletions 1`] = `
"Fetching latest master...
Computing coverage on diff...
No diff information found.


No coverage information found for any of the new lines in the diff."
`;

exports[`should not report if nothing changes 1`] = `
"Fetching latest master...
Computing coverage on diff...
No diff information found.


No coverage information found for any of the new lines in the diff."
`;

exports[`should report all three 1`] = `
"Fetching latest master...
Computing coverage on diff...
Diff files:

  tmp/reports/lcov.info
  util.ts

Diff coverage results:

  F = Function coverage
  L = Line coverage
  B = Branch coverage

------------------------------------------------------------------------------
util.ts
------------------------------------------------------------------------------
1   F | L | B(1/2)export const add = (num1: number, num2: number) => num1 ? num1 + num2 : num1
2                 export const subtract = (num1: number, num2: number) => num1 - num2
3   F | L         export const divide = (num1: number, num2: number) => num1 / num2
4   L             export const math = { add, divide, subtract }
------------------------------------------------------------------------------


Diff coverage statistics:
------------------------------------------------------------------------------
|     lines      |    branches    |   functions    |   path
------------------------------------------------------------------------------
|  33% (1/3)     |  50% (1/2)     |  50% (1/2)     | util.ts
------------------------------------------------------------------------------

Total diff coverage:

 - lines: 33% (1/3)
 - branches: 50% (1/2)
 - functions: 50% (1/2)

ERROR: Insufficient line coverage of 33% (1/3). Required 90%.
ERROR: Insufficient branch coverage of 50% (1/2). Required 80%.
ERROR: Insufficient function coverage of 50% (1/2). Required 90%."
`;

exports[`should report changes across multiple files 1`] = `
"Fetching latest master...
Computing coverage on diff...
Diff files:

  tmp/reports/lcov.info
  util.ts
  util1.ts

Diff coverage results:

  F = Function coverage
  L = Line coverage
  B = Branch coverage

------------------------------------------------------------------------------
util.ts
------------------------------------------------------------------------------
1   F | L | B(1/2)export const add = (num1: number, num2: number) => num1 ? num1 + num2 : num1
2                 export const subtract = (num1: number, num2: number) => num1 - num2
3   F | L         export const divide = (num1: number, num2: number) => num1 / num2
4   L             export const math = { add, divide, subtract }
------------------------------------------------------------------------------

------------------------------------------------------------------------------
util1.ts
------------------------------------------------------------------------------
1   F | L | B(0/2)export const add = (num1: number, num2: number) => num1 ? num1 + num2 : num1
2   F | L         export const subtract = (num1: number, num2: number) => num1 - num2
3   F | L         export const divide = (num1: number, num2: number) => num1 / num2
4   L             export const math = { add, divide, subtract }
------------------------------------------------------------------------------


Diff coverage statistics:
------------------------------------------------------------------------------
|     lines      |    branches    |   functions    |   path
------------------------------------------------------------------------------
|  33% (1/3)     |  50% (1/2)     |  50% (1/2)     | util.ts
|   0% (0/4)     |   0% (0/2)     |   0% (0/3)     | util1.ts
------------------------------------------------------------------------------

Total diff coverage:

 - lines: 14% (1/7)
 - branches: 25% (1/4)
 - functions: 20% (1/5)

ERROR: Insufficient line coverage of 14% (1/7). Required 90%.
ERROR: Insufficient branch coverage of 25% (1/4). Required 80%.
ERROR: Insufficient function coverage of 20% (1/5). Required 90%."
`;

exports[`should report new branches 1`] = `
"Fetching latest master...
Computing coverage on diff...
Diff files:

  tmp/reports/lcov.info
  util.ts

Diff coverage results:

  F = Function coverage
  L = Line coverage
  B = Branch coverage

------------------------------------------------------------------------------
util.ts
------------------------------------------------------------------------------
1   F | L | B(1/2)export const add = (num1: number, num2: number) => num1 ? num1 + num2 : num1
2                 export const subtract = (num1: number, num2: number) => num1 - num2
------------------------------------------------------------------------------


Diff coverage statistics:
------------------------------------------------------------------------------
|     lines      |    branches    |   functions    |   path
------------------------------------------------------------------------------
| 100% (1/1)     |  50% (1/2)     | 100% (1/1)     | util.ts
------------------------------------------------------------------------------

Total diff coverage:

 - lines: 100% (1/1)
 - branches: 50% (1/2)
 - functions: 100% (1/1)

ERROR: Insufficient branch coverage of 50% (1/2). Required 80%."
`;

exports[`should report new functions 1`] = `
"Fetching latest master...
Computing coverage on diff...
Diff files:

  tmp/reports/lcov.info
  util.ts

Diff coverage results:

  F = Function coverage
  L = Line coverage
  B = Branch coverage

------------------------------------------------------------------------------
util.ts
------------------------------------------------------------------------------
1                 export const add = (num1: number, num2: number) => num1 + num2
2                 export const subtract = (num1: number, num2: number) => num1 - num2
3   F | L         export const divide = (num1: number, num2: number) => num1 / num2
------------------------------------------------------------------------------


Diff coverage statistics:
------------------------------------------------------------------------------
|     lines      |    branches    |   functions    |   path
------------------------------------------------------------------------------
| 100% (1/1)     | 100% (0/0)     |   0% (0/1)     | util.ts
------------------------------------------------------------------------------

Total diff coverage:

 - lines: 100% (1/1)
 - branches: 100% (0/0)
 - functions: 0% (0/1)

ERROR: Insufficient function coverage of 0% (0/1). Required 90%."
`;

exports[`should report new lines 1`] = `
"Fetching latest master...
Computing coverage on diff...
No diff information found.


No coverage information found for any of the new lines in the diff."
`;
